x/ibc/core/02-client/keeper/keeper.go

```diff
@@ -125,6 +125,49 @@ func (k Keeper) GetAllGenesisClients(ctx sdk.Context) types.IdentifiedClientStat
 	return genClients.Sort()
 }
 
+// GetAllClientMetadata will take a list of IdentifiedClientState and return a list
+// of IdentifiedGenesisMetadata necessary for exporting and importing client metadata
+// into the client store.
+func (k Keeper) GetAllClientMetadata(ctx sdk.Context, genClients []types.IdentifiedClientState) ([]types.IdentifiedGenesisMetadata, error) {
+	genMetadata := make([]types.IdentifiedGenesisMetadata, 0)
+	for _, ic := range genClients {
+		cs, ok := ic.ClientState.GetCachedValue().(exported.ClientState)
+		if !ok {
+			return nil, sdkerrors.Wrapf(types.ErrInvalidClient, "client state is not exported.ClientState")
+		}
+		gms := cs.ExportMetadata(k.ClientStore(ctx, ic.ClientId))
+		if len(gms) == 0 {
+			continue
+		}
+		clientMetadata := make([]types.GenesisMetadata, len(gms))
+		for i, metadata := range gms {
+			cmd, ok := metadata.(types.GenesisMetadata)
+			if !ok {
+				return nil, sdkerrors.Wrapf(types.ErrInvalidClientMetadata, "expected metadata type: %T, got: %T",
+					types.GenesisMetadata{}, cmd)
+			}
+			clientMetadata[i] = cmd
+		}
+		genMetadata = append(genMetadata, types.NewIdentifiedGenesisMetadata(
+			ic.ClientId,
+			clientMetadata,
+		))
+	}
+	return genMetadata, nil
+}
+
+// SetAllClientMetadata takes a list of IdentifiedGenesisMetadata and stores all of the metadata in the client store at the appropriate paths.
+func (k Keeper) SetAllClientMetadata(ctx sdk.Context, genMetadata []types.IdentifiedGenesisMetadata) {
+	for _, igm := range genMetadata {
+		// create client store
+		store := k.ClientStore(ctx, igm.ClientId)
+		// set all metadata kv pairs in client store
+		for _, md := range igm.ClientMetadata {
+			store.Set(md.GetKey(), md.GetValue())
```

Comment:
should we be validating that it doesn't accidentally overwrite its own client/consensus states?

Reply:
I think you could just do this in genesis validate basic checking that the key isn't == to clientState/consensusStates. In general not much we can do here since it could overwrite it's own data during runtime

Reply:
Hmm I don't see a big use in this. If they export the clientState/consensusState as well, it's not unsafe because they would just overwrite the clients/consensus with the exact same values again which isn't wrong, it's just inefficient.

They would have to do something perversely wrong, to overwrite with something else. But as you said, this is something they already have the power to do. We have to make the assumption that client implementations do not incorrectly write to their own client stores.

Reply:
I think it would be beneficial to do key validation as well. It's not a big deal as it will only be done during the genesis initialization

Reply:
because of the ordering, if the metadata has the `clientState` key it would overwrite the client state which would probably cause a panic (client state path exists but can't unmarshal to client state). Key validation would just prevent a chain from accidentally doing this (note this has nothing to do with the implementation since this is on init genesis)

Reply:
alternatively we could write the metadata first so that the client state and consensus state overwrite metadata if necessary

Reply:
> alternatively we could write the metadata first so that the client state and consensus state overwrite metadata if necessary

Decided to go with this approach. Doing the former would require adding new helper functions and testing (and reviewing) them for correctness. This is a quicker fix that does the same thing and we can choose a more elegant solution after stargate if we wish. I think we will want to make the IBC genesis code a lot more robust